operator/pkg/ingress/ingress.go

```diff
@@ -0,0 +1,524 @@
+		break
+	case ingressServiceUpdatedEvent:
+		log.WithField("service", ev.ingressService.Name).Info("Handling ingress service updated event")
+		err = ic.handleIngressServiceUpdatedEvent(ev)
+		break
+	default:
+		err = fmt.Errorf("received an unknown event: %s", ev)
+	}
+	return err
+}
+
+func (ic *IngressController) getByKey(key string) (*slim_networkingv1.Ingress, error) {
+	objFromCache, exists, err := ic.ingressStore.GetByKey(key)
+	if err != nil {
+		log.WithError(err).Warn("Ingress cache lookup failed")
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("ingress '%s' not found", key)
+	}
+	ingress, ok := objFromCache.(*slim_networkingv1.Ingress)
+	if !ok {
+		return nil, fmt.Errorf("got an invalid object from ingress cache")
+	}
+	return ingress, nil
+}
+
+func getServiceNameForIngress(ingress *slim_networkingv1.Ingress) string {
+	return ciliumIngressPrefix + ingress.Name
+}
+
+func getIngressKeyForService(service *slim_corev1.Service) string {
+	ingressName := strings.TrimPrefix(service.Name, ciliumIngressPrefix)
+	return fmt.Sprintf("%s/%s", service.Namespace, ingressName)
+}
+
+func (ic *IngressController) createLoadBalancer(ingress *slim_networkingv1.Ingress) error {
+	svc := getServiceForIngress(ingress)
+	svcKey, err := cache.MetaNamespaceKeyFunc(svc)
+	if err != nil {
+		log.Warn("failed to get service key for ingress")
+		return err
+	}
+
+	_, exists, err := ic.serviceManager.getByKey(svcKey)
+	if err != nil {
+		log.WithError(err).Warn("service lookup returned an error")
+		return err
+	}
+	if exists {
+		// Service already exists in the cache. For now assume that it was created by the ingress
+		// controller.
+		log.WithField("service", svcKey).Info("Service already exists. Continuing...")
+		return nil
+	}
+
+	_, err = k8s.Client().CoreV1().Services(ingress.Namespace).Create(context.Background(), svc, metav1.CreateOptions{})
+	if err != nil {
+		log.WithError(err).WithField("ingress", ingress.Name).Error("Failed to create a service for ingress")
+	}
+	log.WithField("service", svcKey).Info("Created Service for Ingress")
+
+	return err
+}
+
+func (ic *IngressController) createEndpoints(ingress *slim_networkingv1.Ingress) error {
+	endpoints := getEndpointsForIngress(ingress)
+	key, err := cache.MetaNamespaceKeyFunc(endpoints)
+	if err != nil {
+		log.WithError(err).Warn("MetaNamespaceKeyFunc returned an error")
+		return err
+	}
+
+	// check if the endpoints resource already exists
+	_, exists, err := ic.endpointManager.getByKey(key)
+	if err != nil {
+		log.WithError(err).Warn("endpoints lookup returned an error")
+		return err
+	}
+	if exists {
+		// Endpoints already exists in the cache. For now assume that it was created by the ingress
+		// controller.
+		log.WithField("endpoints", key).Info("Endpoints already exists. Continuing...")
+		return nil
+	}
+
+	_, err = k8s.Client().CoreV1().Endpoints(ingress.Namespace).Create(context.Background(), endpoints, metav1.CreateOptions{})
```

Comment:
The same comment I have left for Endpoints, the same I'm leaving for EndpointSlices. We need to make sure EndpointSlices are also handled.

Reply:
we can use the above thread for discussing this point.